pkg/kubernetes/container/v1alpha1/container.go

```diff
@@ -0,0 +1,194 @@
+type Predicate func(*container) (nameOrMsg string, ok bool)
+
+// predicateFailedError returns the provided predicate as an error
+func predicateFailedError(message string) error {
+	return errors.Errorf("predicatefailed: %s", message)
+}
+
+var (
+	validationFailedError = errors.New("container validation failed")
+)
+
+// asContainer transforms this container instance into corresponding kubernetes
+// container type
+func (c *container) asContainer() corev1.Container {
+	return corev1.Container{
+		Name:                     c.Name,
+		Image:                    c.Image,
+		Command:                  c.Command,
+		Args:                     c.Args,
+		WorkingDir:               c.WorkingDir,
+		Ports:                    c.Ports,
+		EnvFrom:                  c.EnvFrom,
+		Env:                      c.Env,
+		Resources:                c.Resources,
+		VolumeMounts:             c.VolumeMounts,
+		VolumeDevices:            c.VolumeDevices,
+		LivenessProbe:            c.LivenessProbe,
+		ReadinessProbe:           c.ReadinessProbe,
+		Lifecycle:                c.Lifecycle,
+		TerminationMessagePath:   c.TerminationMessagePath,
+		TerminationMessagePolicy: c.TerminationMessagePolicy,
+		ImagePullPolicy:          c.ImagePullPolicy,
+		SecurityContext:          c.SecurityContext,
+		Stdin:                    c.Stdin,
+		StdinOnce:                c.StdinOnce,
+		TTY:                      c.TTY,
+	}
+}
+
+// New returns a new kubernetes container
+func New(opts ...OptionFunc) corev1.Container {
+	c := &container{}
+	for _, o := range opts {
+		o(c)
+	}
+	return c.asContainer()
+}
+
+// builder provides utilities required to build a kubernetes container type
+type builder struct {
+	con    *container  // container instance
+	checks []Predicate // validations to be done while building the container instance
+	errors []error     // errors found while building the container instance
+}
+
+// Builder returns a new instance of builder
+func Builder() *builder {
+	return &builder{
+		con: &container{},
+	}
+}
+
+// validate will run checks against container instance
+func (b *builder) validate() error {
+	for _, c := range b.checks {
+		if m, ok := c(b.con); !ok {
+			b.errors = append(b.errors, predicateFailedError(m))
+		}
+	}
+	if len(b.errors) == 0 {
+		return nil
+	}
+	return validationFailedError
+}
+
+// Build returns the final kubernetes container
+func (b *builder) Build() (corev1.Container, error) {
+	err := b.validate()
+	if err != nil {
+		return corev1.Container{}, err
+	}
+	return b.con.asContainer(), nil
+}
+
+// AddCheck adds the predicate as a condition to be validated against the
+// container instance
+func (b *builder) AddCheck(p Predicate) *builder {
+	b.checks = append(b.checks, p)
+	return b
+}
+
+// AddChecks adds the provided predicates as conditions to be validated against
+// the container instance
+func (b *builder) AddChecks(p []Predicate) *builder {
+	for _, check := range p {
+		b.AddCheck(check)
+	}
+	return b
+}
+
```

Comment:
Why is the additional `AddCheck` function required ? IMO -- we can have following
```
func (b *builder) AddChecks(p []Predicate) *builder {
	for _, check := range p {
		b.checks = append(b.checks, p)
	}
	return b
}
```
Similarly, at other places.